### DISCUSSION TITLE
Why the DOM tree is such complicated?

### DISCUSSION THREAD

**Discussion by cyfung1031** [PrismJS/prism#3013] [Q&A] [answered]
I know that the `tokens` are consists of multiple layers.
However, when it is converted to HTML code, it can be flatten.

### Complicated DOM Tree ###
![ascdqweqwdd](https://user-images.githubusercontent.com/44498510/127364489-708739b4-7556-4cb1-941d-e18bc91a7847.png)

### Flatten DOM Tree ###
![awqdscsd](https://user-images.githubusercontent.com/44498510/127364478-13be72cd-7c69-4bdf-839f-2313013814b4.png)

### As shown above, there is no visual difference between Flatten  ###

In other highlighting libraries, each line are wrapped, but PrismJS does not.
I guess it is to flatten the DOM tree to make it faster.
However, I also see such complicated DOM Tree which is not usual in other highlighting libraries.

### one layer of `span` for making texts ###
![sdcsd](https://user-images.githubusercontent.com/44498510/127365284-7e436433-1829-476a-9ae2-67b46910ca37.PNG)

**Comment by RunDevelopment** [comment, marked as answer]
The nested DOM tree is a byproduct of Prism's recursive highlighting algorithm. The DOM structure just reflects the language definitions (e.g. [Markup tag](https://github.com/PrismJS/prism/blob/ce5e0f014e85b98e9471ab0f89e087d69940c4ac/components/prism-markup.js#L28) is nested, so the DOM tree it produces also has that nesting).

This is also quite useful because the CSS themes have access to (simple) structural information about the code. This allows for very specific custom highlighting. You won't see this kind of highlighting with Prism's standard themes (they are all quite simple) but we have users that do that.

Regarding the lines: Many syntax highlighters are line-based (e.g. all TextMate-based syntax highlighters) but Prism is not. Prism highlights the whole text in one go. Splitting everything into lines is actually quite challenging.

That being said, a plugin that splits everything into lines is being worked on and will be part of Prism v2.
